# Caeser cipher 

In [1]:
import math

math.log(26) / math.log(2)

4.700439718141093

## continuous order substitution

Plaintext Space : $26=2^{4.70}$

Key Space : $26=2^{4.70}$

Example Key:

| Original | a | b | c | d | e | f | g | h | i | j | k | l | m | n | o | p | q | r | s | t | u | v | w | x | y | z |
|----------|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|
| Cipher   | t | u | v | w | x | y | z | a | b | c | d | e | f | g | h | i | j | k | l | m | n | o | p | q | r | s |

issue : the key space too small


**Any secure encryption scheme must have a key space that is not
vulnerable to exhaustive search.**

## Mono-alphabetic substitution

Plaintext Space : $26=2^{4.70}$

Key Space : $26!\approx2^{88}$

Example Key:

| Original | a | b | c | d | e | f | g | h | i | j | k | l | m | n | o | p | q | r | s | t | u | v | w | x | y | z |
|----------|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|
| Cipher   | x | e | u | a | d | n | b | k | v | m | r | o | c | q | f | s | y | h | w | g | l | z | i | j | p | t |

Attact example

Average letter frequencies in the English language:

| a | b | c | d | e | f | g | h | i | j | k | l | m | n | o | p | q | r | s | t | u | v | w | x | y | z |
|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|
| 8.2 | 1.5 | 2.8 | 4.2 | 12.7 | 2.2 | 2.0 | 6.1 | 7.0 | 0.1 | 0.8 | 4.0 | 2.4 | 6.7 | 7.5 | 1.9 | 0.1 | 6.0 | 6.3 | 9.0 | 2.8 | 1.0 | 2.4 | 2.0 | 0.1 | 0.1 |

plaintext:

weconcludethatalthoughthemonoalphabeticcipherh
asaverylargekeyspaceitisstillcompletelyinsecur
ethisisanotherimportantlessonnamelyalthoughala
rgekeyspaceisnecessaryforanysecurecipheritisve
ryfarfrombeingsucient

ciphertext:

idufquoladgkxgxogkflbkgkdcfqfxoskxedgvuuvskdhk
xwxzdhpoxhbdrdpwsxudvgvwwgvooufcsodgdopvqwdulh
dgkvwvwxqfgkdhvcsfhgxqgodwwfqqxcdopxogkflbkxox
hbdrdpwsxudvwqdudwwxhpnfhxqpwdulhduvskdhvgvwzd
hpnxhnhfcedvqbwluvdqg

In [12]:
import numpy as np

known_distribute =np.array([
    ('a', 8.2), ('b', 1.5), ('c', 2.8), ('d', 4.2), ('e', 12.7), ('f', 2.2), 
    ('g', 2.0), ('h', 6.1), ('i', 7.0), ('j', 0.1), ('k', 0.8), ('l', 4.0), 
    ('m', 2.4), ('n', 6.7), ('o', 7.5), ('p', 1.9), ('q', 0.1), ('r', 6.0), 
    ('s', 6.3), ('t', 9.0), ('u', 2.8), ('v', 1.0), ('w', 2.4), ('x', 2.0), 
    ('y', 0.1), ('z', 0.1)
])
known_distribute

array([['a', '8.2'],
       ['b', '1.5'],
       ['c', '2.8'],
       ['d', '4.2'],
       ['e', '12.7'],
       ['f', '2.2'],
       ['g', '2.0'],
       ['h', '6.1'],
       ['i', '7.0'],
       ['j', '0.1'],
       ['k', '0.8'],
       ['l', '4.0'],
       ['m', '2.4'],
       ['n', '6.7'],
       ['o', '7.5'],
       ['p', '1.9'],
       ['q', '0.1'],
       ['r', '6.0'],
       ['s', '6.3'],
       ['t', '9.0'],
       ['u', '2.8'],
       ['v', '1.0'],
       ['w', '2.4'],
       ['x', '2.0'],
       ['y', '0.1'],
       ['z', '0.1']], dtype='<U32')

In [37]:
from collections import Counter

text = """
We conclude that, although the mono-alphabetic cipher has a very large
key space, it is still completely insecure. This is another important lesson.
Namely, although a large key space is necessary for any secure cipher, it is
very far from being sucient.
"""
# Convert the text to lowercase and filter out non-letter characters
plaintext = [char for char in text.lower() if char.isalpha()]
print("".join(plaintext))
key = {
    "a": "x",
    "b": "e",
    "c": "u",
    "d": "a",
    "e": "d",
    "f": "n",
    "g": "b",
    "h": "k",
    "i": "v",
    "j": "m",
    "k": "r",
    "l": "o",
    "m": "c",
    "n": "q",
    "o": "f",
    "p": "s",
    "q": "y",
    "r": "h",
    "s": "w",
    "t": "g",
    "u": "l",
    "v": "z",
    "w": "i",
    "x": "j",
    "y": "p",
    "z": "t",
}

letter_counts = Counter(plaintext)

known_distribute_list = []
for letter, count in letter_counts.items():
    known_distribute_list.append((letter, count / len(plaintext) * 100))

known_distribute = np.array(known_distribute_list)

known_distribute

weconcludethatalthoughthemonoalphabeticcipherhasaverylargekeyspaceitisstillcompletelyinsecurethisisanotherimportantlessonnamelyalthoughalargekeyspaceisnecessaryforanysecurecipheritisveryfarfrombeingsucient


array([['w', '0.4878048780487805'],
       ['e', '13.170731707317074'],
       ['c', '5.853658536585367'],
       ['o', '5.365853658536586'],
       ['n', '5.365853658536586'],
       ['l', '5.853658536585367'],
       ['u', '2.9268292682926833'],
       ['d', '0.4878048780487805'],
       ['t', '7.317073170731707'],
       ['h', '5.853658536585367'],
       ['a', '8.780487804878048'],
       ['g', '2.4390243902439024'],
       ['m', '2.4390243902439024'],
       ['p', '3.414634146341464'],
       ['b', '0.975609756097561'],
       ['i', '7.317073170731707'],
       ['r', '6.829268292682928'],
       ['s', '7.804878048780488'],
       ['v', '0.975609756097561'],
       ['y', '3.902439024390244'],
       ['k', '0.975609756097561'],
       ['f', '1.4634146341463417']], dtype='<U32')

In [36]:
from collections import Counter

ciphertext = [key[char] for char in plaintext]
print("".join(ciphertext))
letter_counts = Counter(ciphertext)

statistics_distribute_list = []
for letter, count in letter_counts.items():
    statistics_distribute_list.append((letter, count / len(ciphertext) * 100))

statistics_distribute = np.array(statistics_distribute_list)

statistics_distribute

idufquoladgkxgxogkflbkgkdcfqfxoskxedgvuuvskdhkxwxzdhpoxhbdrdpwsxudvgvwwgvooufcsodgdopvqwdulhdgkvwvwxqfgkdhvcsfhgxqgodwwfqqxcdopxogkflbkxoxhbdrdpwsxudvwqdudwwxhpnfhxqpwdulhduvskdhvgvwzdhpnxhnhfcedvqbwluvdqg


array([['i', '0.4878048780487805'],
       ['d', '13.170731707317074'],
       ['u', '5.853658536585367'],
       ['f', '5.365853658536586'],
       ['q', '5.365853658536586'],
       ['o', '5.853658536585367'],
       ['l', '2.9268292682926833'],
       ['a', '0.4878048780487805'],
       ['g', '7.317073170731707'],
       ['k', '5.853658536585367'],
       ['x', '8.780487804878048'],
       ['b', '2.4390243902439024'],
       ['c', '2.4390243902439024'],
       ['s', '3.414634146341464'],
       ['e', '0.975609756097561'],
       ['v', '7.317073170731707'],
       ['h', '6.829268292682928'],
       ['w', '7.804878048780488'],
       ['z', '0.975609756097561'],
       ['p', '3.902439024390244'],
       ['r', '0.975609756097561'],
       ['n', '1.4634146341463417']], dtype='<U32')

In [33]:
# Assuming known_distribute and statistics_distribute are defined

# Sort both arrays based on the second element of each tuple
known_distribute = known_distribute[known_distribute[:,1].argsort()]
statistics_distribute = statistics_distribute[statistics_distribute[:,1].argsort()]

# Create a new array of tuples
combined = np.array(list(zip(statistics_distribute[:,0],known_distribute[:,0] )))
combined

array([['i', 'w'],
       ['a', 'd'],
       ['z', 'v'],
       ['e', 'b'],
       ['r', 'k'],
       ['n', 'f'],
       ['d', 'e'],
       ['b', 'g'],
       ['c', 'm'],
       ['l', 'u'],
       ['s', 'p'],
       ['p', 'y'],
       ['q', 'n'],
       ['f', 'o'],
       ['k', 'h'],
       ['u', 'c'],
       ['o', 'l'],
       ['h', 'r'],
       ['g', 't'],
       ['v', 'i'],
       ['w', 's'],
       ['x', 'a']], dtype='<U1')

In [34]:
# Create a mapping dictionary
mapping = {original: mapped for original, mapped in combined}

mapped_ciphertext = "".join(mapping.get(letter, letter) for letter in ciphertext)

print(mapped_ciphertext)

weconcludethatalthoughthemonoalphabeticcipherhasaverylargekeyspaceitisstillcompletelyinsecurethisisanotherimportantlessonnamelyalthoughalargekeyspaceisnecessaryforanysecurecipheritisveryfarfrombeingsucient
